### PR TITLE
Add Row Json Deserializer

### PR DESCRIPTION
**Pull request by akedin**
Add basic Deserializer to allow converting Json objects to Rows.
Doesn't support nullables at the moment. Isn't wired up to SQL yet.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by akedin** [comment]
R: @apilloud @kennknowles

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java [hunk 1]
```diff
@@ -238,17 +238,35 @@ public FieldType type() {
    */
   @AutoValue
   public abstract static class FieldType implements Serializable {
-    // Returns the type of this field.
```
Comment (apilloud) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java]: All these comment changes seem unrelated to a Json Deserialzier. Separate commit?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+      JsonNode jsonNode,
+      Schema targetRowSchema) {
+    return
+        IntStream
+            .range(0, targetRowSchema.getFieldCount())
+            .mapToObj(i -> getFieldValue(jsonNode, targetRowSchema, i))
+            .collect(Row.toRow(targetRowSchema));
+  }
+
+  private static Object getFieldValue(
```
Comment (apilloud) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java]: nit: Following this is kind of confusing. It seems like this function should really part of your `mapToObj` call above.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+      Schema.FieldType currentFieldType) {
+
+    Schema.TypeName fieldTypeName = currentFieldType.getTypeName();
+
+    if (currentValueNode == null) {
+      throw new UnsupportedRowJsonException(
+          "Field " + currentFieldName + " is not present in the JSON object");
+    }
+
+    if (currentValueNode.isNull()) {
```
Comment (apilloud) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java]: I didn't see a test that covered null values. Might be worth testing.

**Comment by akedin** [comment]
@kennknowles:
> 1. It might be preferable to have the conversion driven by the expected type in the schema instead of by the JSON type. It doesn't seem to be a problem, though.

That's what happens. [These checks](https://github.com/apache/beam/pull/5089/files#diff-cbf1f27a95b38256510d5da5e1ae892eR118) look at the schema type, and [primitive value look up](https://github.com/apache/beam/pull/5089/files#diff-cbf1f27a95b38256510d5da5e1ae892eR175) also happens by schema type name.

@kennknowles 
> Can you have tests for out-of-bounds numerical constants? I may have just missed them. The reason I ask is to deliberately focus on the places where the JSON is valid and almost fits the schema, but there is a minor mismatch in the data models.

Working

**Comment by akedin** [comment]
@kennknowles added test for [overflowing integer numbers](https://github.com/apache/beam/pull/5089/files#diff-49296389618a7386e586e41f6d1bff66R377)

**Comment by kennknowles** [comment]
Ah, sorry I misread that. Then the choice of when to reject and when to wrap and when to lose precision is a bit complicated. I think we might need a design doc and to discuss a little bit on dev@ as a next gen of the schema design, but we can move forward with almost anything for now.

**Comment by akedin** [comment]
@kennknowles  added stricter validation, does this approach look better?

Few quirks:

 - not sure if I can quickly come up with a prettier solution for validation, JSON+Jackson+Java makes a weird combination;
 - jackson allows few type widening [configuration options](https://github.com/FasterXML/jackson-databind/wiki/Deserialization-Features), like it can automatically convert all ints to `BigIntegers`, I haven't tested that bit yet. And it's an `ObjectMapper` configuration option, so it's not directly exposed to the serializer. And it doesn't feel right to automatically widen everything;
 - jackson always parses `1.02e5` as `float` node but `1.02` as `double` node, which leads to things like `1.02d != (double)(float)(1.02d)`. Easy workaround is to define Schema field as `Long` but it feels super subtle and unclear if you don't have a debugger at hand. I don't believe there's a configuration flag for this;
 - additionally, rejecting `1.00 -> int` conversion or `number -> string` feels restrictive;

At this point I think that having a more restricting approach is better even while we don't have a robust type system in Schemas:
 - probably easier to make it less restricting in general when we encounter the use cases which require less strict conversions;
 - configuration options to control strictness can be wired up easier when there's something to wire them up to;

**Comment by akedin** [comment]
run java gradle precommit

**Comment by akedin** [comment]
run java precommit

**Comment by akedin** [comment]
run java precommit

**Comment by kennknowles** [comment]
Yes, I like this. If we do decide to make the extractors less strict, the structure of things can stay the same.
